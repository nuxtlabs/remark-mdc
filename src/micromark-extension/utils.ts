import type { Effects } from 'micromark-factory-space'

// Measure the number of character codes in chunks.
// Counts tabs based on their expanded size, and CR+LF as one character.
export function sizeChunks (chunks: any[]) {
  let index = -1
  let size = 0

  while (++index < chunks.length) {
    size += typeof chunks[index] === 'string' ? chunks[index].length : 1
  }

  return size
}

export function prefixSize (events: any, type: string) {
  const tail = events[events.length - 1]
  if (!tail || tail[1].type !== type) { return 0 }
  return sizeChunks(tail[2].sliceStream(tail[1]))
}

/**
 * Calculate line indention size, line indention could be consists of multiple `linePrefix` events
 * @param events parser tokens
 * @returns line indention size
 */
export function linePrefixSize (events: any[]) {
  let size = 0
  let index = events.length - 1
  let tail = events[index]
  while (index >= 0 && tail && tail[1].type === 'linePrefix' && tail[0] === 'exit') {
    size += sizeChunks(tail[2].sliceStream(tail[1]))
    index -= 1
    tail = events[index]
  }

  return size
}

/**
 * Manage token state
 */
export const useTokenState = (tokenName: string) => {
  const token = {
    isOpen: false,
    /**
     * Enter into token, close previous open token if any
     */
    enter: (effects: Effects) => {
      const initialState = token.isOpen
      token.exit(effects)
      effects.enter(tokenName)
      token.isOpen = true

      // Revert to initial state
      return () => {
        token.isOpen = initialState
      }
    },
    /**
     * Enter into token only once, if token is already open, do nothing
     */
    enterOnce: (effects: Effects) => {
      const initialState = token.isOpen

      if (!token.isOpen) {
        effects.enter(tokenName)
        token.isOpen = true
      }

      // Revert to initial state
      return () => {
        token.isOpen = initialState
      }
    },
    /**
     * Exit from token if it is open
     */
    exit: (effects: Effects) => {
      const initialState = token.isOpen
      if (token.isOpen) {
        effects.exit(tokenName)
        token.isOpen = false
      }

      // Revert to initial state
      return () => {
        token.isOpen = initialState
      }
    }
  }
  return token
}
