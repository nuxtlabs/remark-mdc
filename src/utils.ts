export const NON_UNWRAPABLE_TYPES = [
  'componentContainerSection',
  'componentContainerDataSection',
  'containerComponent',
  'leafComponent'
]
