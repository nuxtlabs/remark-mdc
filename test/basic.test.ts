import { describe } from 'vitest'
import { runMarkdownTests } from './utils'

describe('basic', () => {
  runMarkdownTests({
    simple: {
      mdcOptions: {
        experimental: {
          autoUnwrap: true
        }
      },
      markdown: [
        'paragraph 1',
        '',
        '---',
        '',
        'paragraph 2'
      ].join('\n')
    },
    link: {
      mdcOptions: {
        experimental: {
          autoUnwrap: true
        }
      },
      markdown: [
        '[link](https://nuxtjs.org){target="_blank"}'
      ].join('\n')
    }
  })
})
