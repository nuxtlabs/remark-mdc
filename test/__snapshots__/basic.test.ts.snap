// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic > link 1`] = `
{
  "children": [
    {
      "children": [
        {
          "attributes": {
            "target": "_blank",
          },
          "children": [
            {
              "position": {
                "end": {
                  "column": 6,
                  "line": 1,
                  "offset": 5,
                },
                "start": {
                  "column": 2,
                  "line": 1,
                  "offset": 1,
                },
              },
              "type": "text",
              "value": "link",
            },
          ],
          "position": {
            "end": {
              "column": 27,
              "line": 1,
              "offset": 26,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "title": null,
          "type": "link",
          "url": "https://nuxtjs.org",
        },
      ],
      "position": {
        "end": {
          "column": 44,
          "line": 1,
          "offset": 43,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 44,
      "line": 1,
      "offset": 43,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`basic > simple 1`] = `
{
  "children": [
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 12,
              "line": 1,
              "offset": 11,
            },
            "start": {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "text",
          "value": "paragraph 1",
        },
      ],
      "position": {
        "end": {
          "column": 12,
          "line": 1,
          "offset": 11,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
    {
      "position": {
        "end": {
          "column": 4,
          "line": 3,
          "offset": 16,
        },
        "start": {
          "column": 1,
          "line": 3,
          "offset": 13,
        },
      },
      "type": "thematicBreak",
    },
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 12,
              "line": 5,
              "offset": 29,
            },
            "start": {
              "column": 1,
              "line": 5,
              "offset": 18,
            },
          },
          "type": "text",
          "value": "paragraph 2",
        },
      ],
      "position": {
        "end": {
          "column": 12,
          "line": 5,
          "offset": 29,
        },
        "start": {
          "column": 1,
          "line": 5,
          "offset": 18,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": {
    "end": {
      "column": 12,
      "line": 5,
      "offset": 29,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
